### PR TITLE
Port `remainder` from TH to ATen (CPU and CUDA)

### PR DESCRIPTION
**Pull request by kurtamohler**
CPU issue #24753 
CUDA issue #24615

### CODE REVIEW

**Comment by VitalyFedyunin** [comment]
Please add performance benchmarks CPU/CUDA; Single Thread/Multi Thread

**Comment by kurtamohler** [comment]
CPU performance:

dtype | num_elements | TH CPU single-core time | TH CPU multi-core time | ATen CPU single-core time | ATen CPU multi-core time | Single-core speedup | multi-core speedup
-- | -- | -- | -- | -- | -- | -- | --
torch.float32 | 1000 | 0.0000404 | 0.000037 | 0.000045 | 0.0000484 | 0.8977777778 | 0.7644628099
torch.float32 | 10000 | 0.0001216 | 0.0001082 | 0.0002138 | 0.0002374 | 0.5687558466 | 0.4557708509
torch.float32 | 100000 | 0.0009386 | 0.0004288 | 0.001894 | 0.0008242 | 0.4955649419 | 0.5202620723
torch.float32 | 1000000 | 0.0102662 | 0.0007216 | 0.020459 | 0.0019244 | 0.5017938316 | 0.3749740179
torch.float32 | 10000000 | 0.1067806 | 0.0098368 | 0.2163726 | 0.02088 | 0.4935033364 | 0.4711111111
torch.float64 | 1000 | 0.0000462 | 0.0000418 | 0.000062 | 0.000082 | 0.7451612903 | 0.5097560976
torch.float64 | 10000 | 0.0001254 | 0.0001188 | 0.0003942 | 0.000523 | 0.3181126332 | 0.2271510516
torch.float64 | 100000 | 0.0008038 | 0.0001702 | 0.0037252 | 0.0013866 | 0.2157736497 | 0.1227462859
torch.float64 | 1000000 | 0.0078996 | 0.0004958 | 0.0390346 | 0.0037106 | 0.2023743038 | 0.1336172048
torch.float64 | 10000000 | 0.1048736 | 0.0136762 | 0.3988242 | 0.0384804 | 0.262956962 | 0.3554069085
torch.int32 | 1000 | 0.0000456 | 0.00006 | 0.0000978 | 0.0001258 | 0.4662576687 | 0.4769475358
torch.int32 | 10000 | 0.0001824 | 0.0001844 | 0.000755 | 0.0009836 | 0.241589404 | 0.1874745832
torch.int32 | 100000 | 0.0015858 | 0.0002876 | 0.0073278 | 0.0028862 | 0.2164087448 | 0.09964659414
torch.int32 | 1000000 | 0.0150388 | 0.0008232 | 0.0732916 | 0.006207 | 0.205191318 | 0.1326244563
torch.int32 | 10000000 | 0.1661234 | 0.012534 | 0.7425 | 0.0699036 | 0.2237352189 | 0.1793040702
torch.int64 | 1000 | 0.0000424 | 0.000043 | 0.0000932 | 0.0001218 | 0.4549356223 | 0.3530377668
torch.int64 | 10000 | 0.0001238 | 0.0001258 | 0.0007162 | 0.0009254 | 0.1728567439 | 0.1359412146
torch.int64 | 100000 | 0.0010318 | 0.0002146 | 0.0069764 | 0.002493 | 0.1478986297 | 0.08608102688
torch.int64 | 1000000 | 0.0085434 | 0.0006556 | 0.0704638 | 0.0063514 | 0.1212452351 | 0.103221337
torch.int64 | 10000000 | 0.1148014 | 0.0144832 | 0.7106772 | 0.064825 | 0.1615380372 | 0.2234199769

GPU performance:

dtype | num_elements | TH GPU time | ATen GPU time | Speedup
-- | -- | -- | -- | --
torch.float32 | 1000000 | 0.0000588 | 0.0000466 | 1.261802575
torch.float32 | 10000000 | 0.0000608 | 0.0000474 | 1.282700422
torch.float32 | 100000000 | 0.000063 | 0.0000456 | 1.381578947
torch.float64 | 1000000 | 0.0000572 | 0.0000482 | 1.186721992
torch.float64 | 10000000 | 0.0000566 | 0.0000476 | 1.18907563
torch.float64 | 100000000 | 0.0000608 | 0.000048 | 1.266666667
torch.int32 | 1000000 | 0.0000556 | 0.0000462 | 1.203463203
torch.int32 | 10000000 | 0.0000566 | 0.0000472 | 1.199152542
torch.int32 | 100000000 | 0.0000546 | 0.0000456 | 1.197368421
torch.int64 | 1000000 | 0.0000566 | 0.0000472 | 1.199152542
torch.int64 | 10000000 | 0.0000584 | 0.0000468 | 1.247863248
torch.int64 | 100000000 | 0.0000564 | 0.0000474 | 1.189873418

As you can see, the CPU performance decreases with my changes. My comment in the CPU issue talks about why I think that is, and why I currently think that should be fixed in a separate issue: https://github.com/pytorch/pytorch/issues/24753#issuecomment-593662643 But if someone thinks differently, my mind is open to being changed!

**Comment by kurtamohler** [comment]
The test failure from the Dr CI comment above is caused by incorrect handling of datatypes in my implementation, when the divisor argument is a scalar. It seems that the scalar needs to be converted to match the datatype of the tensor. This was the old behavior:
```
>>> torch.remainder(torch.tensor([4, 7, 0]), 2.55)
tensor([0, 1, 0])
```
This is the behavior with my changes:
```
>>> torch.remainder(torch.tensor([4, 7, 0]), 2.55)
tensor([1.4500, 1.9000, 0.0000])
```
So I will figure out how to fix that.

**Comment by kurtamohler** [comment]
Looks like I'm getting errors for the XLA device tests:
```
Mar 03 21:48:11 FAIL [0.043s]: test_remainder_inplace_negative_tensor_xla_int16 (__main__.TestTensorDeviceOpsXLA)
Mar 03 21:48:11 ----------------------------------------------------------------------
Mar 03 21:48:11 Traceback (most recent call last):
Mar 03 21:48:11   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 198, in instantiated_test
Mar 03 21:48:11     result = test(self, device_arg, dtype)
Mar 03 21:48:11   File "/var/lib/jenkins/workspace/xla/test/../../test/test_torch.py", line 15314, in fn
Mar 03 21:48:11     self.assertEqual(cpu_tensor, device_tensor, prec=precision)
Mar 03 21:48:11   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 336, in assertEqual
Mar 03 21:48:11     return DeviceTypeTestBase.assertEqual(self, x, y, prec, message, allow_inf, **kwargs)
Mar 03 21:48:11   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 872, in assertEqual
Mar 03 21:48:11     assertTensorsEqual(x, y)
Mar 03 21:48:11   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 842, in assertTensorsEqual
Mar 03 21:48:11     self.assertLessEqual(max_err, prec, message)
Mar 03 21:48:11 AssertionError: tensor(9, dtype=torch.int16) not less than or equal to 0.001 : 
```
I'm not sure how to fix that.

File: aten/src/ATen/Declarations.cwrap [hunk 1]
```diff
@@ -657,43 +657,6 @@
...
-    - cname: remainder
-      arguments:
-        - THTensor* self
-        - THTensor* self
-        - real other
-    - cname: cremainder
-      arguments:
-        - THTensor* self
-        - arg: THTensor* self
-          broadcast: other inplace fallback
```
Comment (gchanan) [inline aten/src/ATen/Declarations.cwrap]: is it the case that currently non-broadcast-able but "same number of element" tensors work?

i.e. (2,3) X (3,2)

Comment (kurtamohler) [reply aten/src/ATen/Declarations.cwrap]: No, that didn't work in TH and it doesn't work in my implementation either. They both fail with the same error:
```
>>> torch.ones([2,3]).remainder(torch.ones([3,2]))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The size of tensor a (3) must match the size of tensor b (2) at non-singleton dimension 1
```

File: aten/src/ATen/native/BinaryOps.cpp [hunk 2]
```diff
@@ -222,6 +241,21 @@ Tensor& div_(Tensor& self, Scalar other) {
   return self.div_(wrapped_scalar_tensor(other)); // redispatch!
 }
 
+Tensor remainder(const Tensor& self, Scalar other) {
+  Tensor other_tensor = wrapped_scalar_tensor(other);
+  return native::remainder(self, other_tensor.toType(self.scalar_type()));
```
Comment (gchanan) [inline aten/src/ATen/native/BinaryOps.cpp]: the scalar type conversion looks like it retains the same behavior, but I don't think it's actually what we want long term

(it differs from NumPy, for example, CC @mruberry).

I think we should leave it as is, but could you just put a comment that reads something like:
"FIXME: do normal type promotion a-la NumPy"

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 3]
```diff
@@ -94,6 +94,31 @@ void div_kernel(TensorIterator& iter) {
...
+          r += b;
+        }
+        return r;
+      });
+    });
+  } else {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter.dtype(), "remainder_cpu", [&]() {
+      cpu_kernel_vec(iter,
+        [=](scalar_t a, scalar_t b) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
+          return a - b * at::native::floor_impl(a / b);
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: the old implementation had a check against `b` being 0 and an explicit set to NaN in that case.  Why the change here?  Is this guaranteed to be the same?  What happens in the vectorized case? (does __ubsan_ignore_float_divide_by_zero__ apply there?)

Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: I didn't do an explicit check because the `div` operator doesn't explicitly check either. But I don't know if this is guaranteed to be the same, so I'll add the check back in.

I'm not sure about the vectorized case. I don't know how to make pytorch tensors use vectorized types.

I don't know what `__ubsan_ignore_float_divide_by_zero__` does, I just copied it from `div`.

File: aten/src/ATen/native/BinaryOps.cpp (hunk 2 above)
Comment (kurtamohler) [reply aten/src/ATen/native/BinaryOps.cpp]: Comment added.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 3 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: It uses vectorization if the array is big enough -- I just added some print statements and e.g.:
`torch.randn(2,3).remainder(0)` calls the scalar version and `torch.randn(200, 300).remainder(0)` calls the Vec256 version.

Comment (gchanan) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: ok, I looked into this a bit -- I'm not actually sure if the vectorized divide by zero is undefined or not -- we could test it and see -- just use the example I gave above.

Comment (gchanan) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: also in this case, we can avoid the undefined behavior by checking if b is 0, like TH and THC do.  That might even be faster?   Can you rerun the benchmarks and check?

Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Ah ok, using large tensors worked for me too. I've added explicit float divide by zero checks.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 4]
```diff
@@ -94,6 +94,42 @@ void div_kernel(TensorIterator& iter) {
...
+        },
+        [=](Vec256<scalar_t> a, Vec256<scalar_t> b) {
+          Vec256<scalar_t> r = a - b * (a / b).floor();
+          Vec256<scalar_t> r_nans = b.map([](const scalar_t x) -> scalar_t {
+            return (x == static_cast<scalar_t>(0)) ?
+              std::numeric_limits<scalar_t>::quiet_NaN() :
+              static_cast<scalar_t>(0);
+          });
+          // Adding a NaN to any other number will result in NaN
+          return r + r_nans;
```
Comment (kurtamohler) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: The way I implemented the explicit divide by zero check for Vec256 might not be ideal. It would be better to be able to avoid all the 0 additions that this introduces. It would be nice if Vec256 had a binary input map function, so I could operate on pairs of elements in `a` and `b` directly. But I only see a unary map function available for Vec256.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 5]
```diff
@@ -94,6 +94,42 @@ void div_kernel(TensorIterator& iter) {
...
+        [=](scalar_t a, scalar_t b) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
+          if (b == 0) {
+            return std::numeric_limits<scalar_t>::quiet_NaN();
+          } else {
+            return a - b * at::native::floor_impl(a / b);
+          }
+        },
+        [=](Vec256<scalar_t> a, Vec256<scalar_t> b) {
+          Vec256<scalar_t> r = a - b * (a / b).floor();
+          Vec256<scalar_t> r_nans = b.map([](const scalar_t x) -> scalar_t {
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: I actually don't know if this part is necessary -- I couldn't find any documentation that defines precisely what the intrinsic does.  And this is certainly going to be slower, unlikely the non-vectorized case.

I was just suggesting adding a test for the vectorized path and seeing if the ubsan / asan build complains.  If it doesn't I think the old vectorized code is fine (might want to add a comment though).

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 4 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Oh never mind, I think I found what I need: `at::vec256::map2`.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 6]
```diff
@@ -94,6 +94,31 @@ void div_kernel(TensorIterator& iter) {
   }
 }
 
+void remainder_kernel(TensorIterator& iter) {
+  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
+    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cpu", [&]() {
+      cpu_kernel(iter, [](scalar_t a, scalar_t b) -> scalar_t {
+        TORCH_CHECK(b != 0, "ZeroDivisionError");
```
Comment (gchanan) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: I'm a bit confused about what is going on here.

In my nightly version of PyTorch, the following happens when I have a tensor with remainder 0:
On CPU, I get Floating point exception (core dumped)
On CUDA, I get a tensor filled with 4294967295.

Is that consistent with what you are seeing?  Why the change to make (at least) CUDA an error?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 5 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Oh alright, makes sense. I'll add a test and see what happens.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 4 above)
Comment (gchanan) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: see my comment above, I'm not sure if this is necessary or not.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 6 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Yeah I do see that behavior. Shouldn't we change that though? It seems better to have a catchable exception rather than a core dump. And if the CPU version raises an exception, shouldn't the CUDA version as well?

**Review by ailzhang — CHANGES_REQUESTED** [review]
Hi thanks for the PR! 
I took a look at XLA test failure and I think it's real. 
For example: 
```
In [2]: a = torch.tensor(6)

In [3]: torch.remainder(a, -3)
Out[3]: tensor(0)
```
But it returns `-3` after this patch. CPU and CUDA tests passed since they were changed together in this PR.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 6 above)
Comment (Baranowski) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: Changing the crash into an exception for `div` was considered high pri: https://github.com/pytorch/pytorch/issues/327

I suspect that changing this behaviour for `remainder` (and also `fmod`) is also desirable.

**Comment by kurtamohler** [comment]
@ailzhang , do you have any suggestions for fixing the XLA support? I don't know anything about how operations are dispatched to that device.

**Comment by ailzhang** [comment]
@kurtamohler The XLA failure is not specific to XLA. I put an example in the comment above that this PR changed returned result of CPU and CUDA. If you fix the example above, XLA will pass as well.

**Comment by kurtamohler** [comment]
@ailzhang , oh I see! Sorry, I misunderstood. I'll go ahead and fix that.

**Comment by ailzhang** [comment]
@kurtamohler Nw! Please also add a test case for the example after your fix! ;) We definitely should have added a test case for those edge cases.

**Comment by kurtamohler** [comment]
@ailzhang , yeah there seem to be a fair amount of missing tests for various cases. I will add them today.

File: aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu [hunk 7]
```diff
@@ -64,9 +64,38 @@ void mul_kernel_cuda(TensorIterator& iter) {
   }
 }
 
+void remainder_kernel_cuda(TensorIterator& iter) {
+  if (isIntegralType(iter.dtype(), /*includeBool*/ false)) {
+    AT_DISPATCH_INTEGRAL_TYPES(iter.dtype(), "remainder_cuda", [&]() {
+      using thrust_t = typename ztype_cuda<scalar_t>::thrust_t;
+      gpu_kernel_with_scalars(iter, []GPU_LAMBDA(thrust_t a, thrust_t b) -> thrust_t {
+        CUDA_KERNEL_ASSERT(b != 0);
```
Comment (kurtamohler) [inline aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu]: Hmmm, it seems that this device-side assert doesn't do what I was hoping. I wanted to create a catchable python exception, but that doesn't seem to happen:
```
>>> import torch
>>> a = torch.ones(1000, dtype=torch.int, device='cuda')
>>> b = torch.ones(1000, dtype=torch.int, device='cuda')
>>> b[500]=0
>>> try:
...     r = a.remainder(b)
... except RuntimeError:
...     print("error caught")
... 
>>> /home/kurtamohler/development/pytorch-remainder/aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu:79: lambda [](int, int)->int::operator()(int, int)->int: block: [1,0,0], thread: [61,0,0] Assertion `b != 0` failed.
```
And interestingly, after running that, I can no longer do anything with cuda without quitting and rerunning python:
```
>>> torch.ones(1000, dtype=torch.int, device='cuda')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: CUDA error: device-side assert triggered
```
Notice that this now throws a python exception, but the exception is not thrown during the `remainder` call.
I'm not sure what the best solution would be.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 5 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: The test cases I added for this pass whether or not the explicit divide by 0 check is there, so I removed the check.

**Comment by kurtamohler** [comment]
@ailzhang , the XLA build is working now. Thanks for the help!

File: aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu (hunk 7 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu]: You can't do device side assert and get an assertion for this. You will have to give up on this until we land https://github.com/pytorch/pytorch/pull/26776/

I'd recommend just preserving the old behavior from CUDA for now

Comment (kurtamohler) [reply aten/src/ATen/native/cuda/BinaryArithmeticKernel.cu]: Ah ok, sounds good to me.

**Review by ezyang — APPROVED** [review]

**Comment by kurtamohler** [comment]
@ezyang , i just pushed the commit to remove the CUDA device-side assert. I think it came in right after you approved. Will phabricator pick it up?

**Comment by ezyang** [comment]
yeah we got it

**Comment by zou3519** [comment]
I think this is marked as BC-breaking because torch.remainder(a, b) now throws an error for division by zero on CPU. Previously it was undefined behavior (torch.remainder(torch.tensor(1.0), torch.tensor(0.)) returns tensor(nan) on my machine, but there are reports in this PR that it can segfault as well)
